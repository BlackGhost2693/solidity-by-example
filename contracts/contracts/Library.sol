pragma solidity ^0.5.16;

library MyLib {
    function incrementBy(uint i, uint val) public pure returns (uint) {
        return i + val;
    }

    // Array function to delete element at index and re-organize the array
    // so that their are no gaps between the elements.
    function deleteElement(string[] storage arr, uint index) internal {
        // Move the last element into the place to delete
        arr[index] = arr[arr.length - 1];

        // Remove the last element
        arr.pop();
    }
}

contract TestMyLib {
    using MyLib for uint;
    using MyLib for string[];

    string[] public array = ["foo", "bar"];

    function testIncrementBy() public pure returns (uint) {
        uint i = 10;

        return i.incrementBy(10);
    }

    function testDeleteElement() public {
        array.push("abc");
        array.deleteElement(1);
        // array should now be ["foo", "abc"]
    }
}

// TODO example MyLib.func()
