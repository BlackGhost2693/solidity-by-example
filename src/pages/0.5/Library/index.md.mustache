Libraries are similar to contracts, but they are deployed only once at a

specific address and their code is reused using the `delegatecall`.

```solidity
{{{Library}}}
```
