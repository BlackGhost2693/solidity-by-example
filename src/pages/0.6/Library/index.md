Libraries are similar to contracts, but you can't declare any state variable and
you can't send ether.

A library is embedded into the contract if all library functions are internal.

Otherwise the library must be deployed and then linked before the contract is deployed.

```solidity
{{{Library}}}
```
