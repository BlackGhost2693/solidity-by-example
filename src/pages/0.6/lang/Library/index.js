import React from "react"
import Example from "../../../components/Example"
import html from "./index.html.js"
import { SOL_VERSION } from "../constants"

export default function (props) {
  return (
    <Example
      title="Library"
      version={SOL_VERSION}
      description="Example of how to write and use libraries in your Solidity code"
      html={html}
    />
  )
}
