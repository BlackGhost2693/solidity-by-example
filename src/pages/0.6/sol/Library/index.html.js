// metadata
export const version = "0.6.0"
export const title = "Library"
export const description = "Example of how to write and use libraries in your Solidity code"

const html = `<p>Libraries are similar to contracts, but you can&#39;t declare any state variable and
you can&#39;t send ether.</p>
<p>A library is embedded into the contract if all library functions are internal.</p>
<p>Otherwise the library must be deployed and then linked before the contract is deployed.</p>
<pre><code class="language-solidity">pragma solidity ^0.6.0;

library SafeMath {
    function add(uint x, uint y) internal pure returns (uint) {
        uint z = x + y;
        require(z &gt;= x, "uint overflow");

        return z;
    }
}

contract TestSafeMath {
    using SafeMath for uint;

    uint public MAX_UINT = 2 ** 256 - 1;

    function testAdd(uint x, uint y) public pure returns (uint) {
        return x.add(y);
    }
}

// Array function to delete element at index and re-organize the array
// so that their are no gaps between the elements.
library Array {
    function remove(uint[] storage arr, uint index) public {
        // Move the last element into the place to delete
        arr[index] = arr[arr.length - 1];
        arr.pop();
    }
}

contract TestArray {
    using Array for uint[];

    uint[] public arr;

    function testArrayRemove() public {
        for (uint i = 0; i &lt; 3; i++) {
            arr.push(i);
        }

        arr.remove(1);

        assert(arr.length == 2);
        assert(arr[0] == 0);
        assert(arr[1] == 2);
    }
}
</code></pre>
`

export default html
