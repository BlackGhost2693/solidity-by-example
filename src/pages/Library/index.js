import React from "react"
import Example from "../../components/Example"
import html from "./index.html.js"

export default function(props) {
  return (
    <Example
      title="Library"
      description="Example of how to write and use libraries in your Solidity code"
      html={html}
    />
  )
}
