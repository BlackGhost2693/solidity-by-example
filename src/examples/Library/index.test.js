import React from "react"
import { shallow } from "enzyme"
import Library from "./index"

test("it renders", () => {
  const component = shallow(<Library />)

  expect(component).toMatchSnapshot()
})
