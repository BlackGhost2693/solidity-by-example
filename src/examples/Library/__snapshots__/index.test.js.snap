// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it renders 1`] = `
<Example
  html="<p>Libraries are similar to contracts, but they are deployed only once at a</p>
<p>specific address and their code is reused using the <code>delegatecall</code>.</p>
<pre><code class=\\"language-solidity\\">pragma solidity ^0.5.1;

library MyLib {
    function incrementBy(uint i, uint val) public pure returns (uint) {
        return i + val;
    }

    // Array function to delete element at index and re-organize the array
    // so that their are no gaps between the elements.
    function deleteElement(string[] storage arr, uint index) internal {
        // Move the last element into the place to delete
        arr[index] = arr[arr.length - 1];

        // Remove the last element
        arr.pop();
    }
}

contract TestMyLib {
    using MyLib for uint;
    using MyLib for string[];

    string[] public array = [\\"foo\\", \\"bar\\"];

    function testIncrementBy() public pure returns (uint) {
        uint i = 10;

        return i.incrementBy(10);
    }

    function testDeleteElement() public {
        array.push(\\"abc\\");
        array.deleteElement(1);
        // array should now be [\\"foo\\", \\"abc\\"]
    }
}
</code></pre>
"
  title="Library"
/>
`;
